core/src/main/java/org/apache/calcite/rel/core/Join.java

```diff
@@ -311,6 +347,10 @@ public abstract Join copy(RelTraitSet traitSet, RexNode conditionExpr,
   public JoinInfo analyzeCondition() {
     return joinInfo;
   }
+
+  @Override public ImmutableList<RelHint> getHints() {
+    return hints;
+  }
```

Comment:
What is the behavior after a rule fires? Is the RelNode still keep the hints?

Reply:
If this `RelNode` is copied, new copied node would keep the hints, but if you create a new node, you should pass the hints explicitly.

Reply:
Maybe i should write a design doc, this seems a little complicated.